user/tlu/module/src/TluRawEvent2StdEventConverter.cc

```diff
@@ -30,20 +29,36 @@ bool TluRawEvent2StdEventConverter::Converting(eudaq::EventSPC d1, eudaq::Standa
     d2->SetTag(TLU+stm+"_TRG", std::to_string(d1->GetTriggerN()));
   }
 
-  auto triggersFired = d1->GetTag("TRIGGER" , "NAN");
-  auto finets0 = d1->GetTag("FINE_TS0", "NAN");
-  auto finets1 = d1->GetTag("FINE_TS1", "NAN");
-  auto finets2 = d1->GetTag("FINE_TS2", "NAN");
-  auto finets3 = d1->GetTag("FINE_TS3", "NAN");
-  auto finets4 = d1->GetTag("FINE_TS4", "NAN");
-  auto finets5 = d1->GetTag("FINE_TS5", "NAN");
+  auto triggersFired = std::stoi(d1->GetTag("TRIGGER" , "0"), nullptr, 2); // interpret as binary
+  auto finets0 = std::stoi(d1->GetTag("FINE_TS0", "0"));
+  auto finets1 = std::stoi(d1->GetTag("FINE_TS1", "0"));
+  auto finets2 = std::stoi(d1->GetTag("FINE_TS2", "0"));
+  auto finets3 = std::stoi(d1->GetTag("FINE_TS3", "0"));
+  auto finets4 = std::stoi(d1->GetTag("FINE_TS4", "0"));
+  auto finets5 = std::stoi(d1->GetTag("FINE_TS5", "0"));
 
   // Set times for StdEvent in picoseconds (timestamps provided in nanoseconds):
-  d2->SetTimeBegin(d1->GetTimestampBegin() * 1000);
+  // Note: __builtin_popcount counts the "ones" in a binary, see here:
+  // https://www.geeksforgeeks.org/count-set-bits-in-an-integer/
+  // I.e. add up all timestamps for which the triggersFired bit is 1 and divide by number of active triggers:
+  // Factor 25./32. -> fine timestamp in 781ps binning
+  auto finets = ((finets0 * ((triggersFired >> 0) & 0x1))
+               + (finets1 * ((triggersFired >> 1) & 0x1))
+               + (finets2 * ((triggersFired >> 2) & 0x1))
+               + (finets3 * ((triggersFired >> 3) & 0x1))
+               + (finets4 * ((triggersFired >> 4) & 0x1))
+               + (finets5 * ((triggersFired >> 5) & 0x1)))
+               / __builtin_popcount(triggersFired & 0x3F); // count "ones" in binary
+
+  auto ts = (d1->GetTimestampBegin() * 1000) & 0xFFFFFFFFFFFFFF00 + static_cast<uint64_t>(25. / 32. * finets);
+
+  d2->SetTimeBegin(ts);
   d2->SetTimeEnd(d1->GetTimestampEnd() * 1000);
+  d2->SetTimestamp(ts, d1->GetTimestampEnd(), d1->IsFlagTimestamp());
```

Comment:
line 55 56 could be not necessary  it  the line 57 always override them.

Reply:
I beg to differ:

* `SetTimestamp()` is a member of `eudaq::Event` and sets a timestamp in clock ticks.
* `SetTimeBegin()` and `SetTimeEnd()` are members of `eudaq::StandardEvent` and set absolute times in picoseconds.

The two variables are complementary.